Initial hw1 for mahshid mansouri (mm64)_For_Pull_Request

mahshidmansouri opened the pull request:
Hi Grace, 

Please review this version for code review!

Thanks,
Mahshid

mahshidmansouri committed:
Add files via upload

mahshidmansouri committed:
Add files via upload

calkins7 commented:
Hi Mahshid,

Overall your code was very easy to understand and follow! I liked how everything was structured and how readable each function title was. Below are my comments:

**General**:
- [ ] I think we are supposed to have a `train_gridworld.py` and `train_pendulum.py` file that runs all the methods for each scenario instead of running a separate file for each different method. You could probably copy and paste all your code at the end of each file into the `train_gridworld.py` and `train_pendulum.py` files into these files and import the functions needed from your other files to run each scenario. 

**Policy Iteration**:
- [ ] In line 32 and line 104, you have `while True:` statements. I would recommend changing this to a `for iter in np.arange(max_allowable_iters):` loop, which will allow you to keep the rest of the code the same and will prevent the algorithm from getting stuck in a while loop if something goes wrong
- [ ] The example trajectory and state value function figures you have are printouts in the terminal, which are helpful for the code user but not for communicating results to others. I would recommend making these either print to Latex / other format tables for insertion into a report or convert the data you have into a figure of some form. 
- [ ] You could also consider putting all your gridworld plotter functions in a single file `plotting.py` and importing them to wherever you run the scenarios to make the code slightly easier to read. 

**Value Iteration:**
- [ ] The function `one_step_lookahead` is used by multiple algorithms - you could consider putting in in a separate `helpers.py` file and importing it to PI and VI using `from helpers import one_step_lookahead` so you don't need to maintain two copies of the same method. 
- [ ] Same comments as for PI apply here as well (while loop on line 54)

**SARSA:**
- [ ] the `choose_action` function is applying an epsilon greedy policy, but I don't see where the probability distribution is normalized by the number of actions. I am not sure if you method is mathematically equivalent to the episolon-greedy distribution described on pg 100 of Sutton and Barto
- [ ] It may improve readability to break the trajectory plot out of the SARSA function into a different function (this could also be helpful because you may want to plot different trajectories based on the environment)
- [ ] You could also make a single function for the part where you run multiple episodes in `learning_curve_plotter_for_different_epsilon` and `learning_curve_plotter_for_different_alpha` where you pass in the alpha and epsilon and run the specified number of episodes. 
- [ ] Since you use `TD_0` multiple times, you could make it a single function and import it to the SARSA and Q-learning function files using `from TD_0_file import TD_0_function` 

**Q-learning:**
- [ ] Same comments as SARSA on readability

**Results/Figures:**
- [ ] I don't think your trajectory plots for the gridworld make sense. Based on the example in the textbook, the optimal policy would be to jump to state A'(21) from state A(2) as many times as possible since that yields the highest reward. The fact that your state "stabilizes" later in the episode means the agent is missing out on potential rewards. This may be something either to do with implementation or the functions themselves but I was unable to determine the cause. I would also expect all 4 algorithms to give the same optimal policy and trajectory when trained with the appropriate number of episodes and hyperparameters.
- [ ] I would recommend plotting the learning curves for the pendulum on a log scale on the y axis to make them more readable because of the negative values at the beginning. 
- [ ] The pendulum trajectories with the optimal policy are not actually performing the optimal policy per your screencap of the optimal policy. (I think those figures were for the gridworld anyway based on the dimensionality of the policy)
- [ ] I think you need to run for more episodes of Q-learning and SARSA to get better results.

mahshidmansouri commented:
Hi Grace, 

Thanks for your valuable comments. Here are my responses to your comments: 

**General:** 

Yes, you're right. I have incorporated all the methods for each of the two problems to their corresponding train.py file. 

**Policy Iteration:** 

1. Thanks for the suggestion. It makes sense to me. Since I was able to get results for this assignment without getting stuck in the while loop, I did not change it to a for loop, but I will make sure to do so for the next assignments and any future coding. 

2. Yes, you're right. Those parts were only for a preliminary check of my results and making sure if they make sense. For the final code, I have removed all those lines from my code and made sure I use better plots/visuals to communicate my results to the readers (e.g., I included V(s) vs. state plot as well as a nice table with four arrows showing each of the four actions that the agent can take in the gridworld problem)

3. Thanks for this suggestion. This will help with making my code even more modular.  I moved all my plotting functions into a single script for plotting purposes only. 

**Value Iteration:** 

1. Thanks for this suggestion. I moved the one_step_lookahead function into a separate script called helpers.py as you suggested. 

2. Same response as my response to (1) in the Policy Iteration feedback. 


**SARSA:** 

1. Thanks for catching this! Initially, I had used a different probability distribution inside this function that needed to be changed, but I forgot to do so. I changed the epsilon-greedy policy function to agree with what we had learned in the course.

2. Thanks for the suggestion. I moved all the plotting functions for SARSA to a separate script called plotters.py to add more modularity to my code. 

3. Thanks for this point. I did not change any of my current code structure as I had already written the plotting functions for varying epsilon and alpha, but I will make sure to follow what you suggested for any future coding assignment. 

4. Thanks for the suggestion. I moved the TD(0) function to a separate helpers.py script.


**Q-Learning** 

Same responses as my responses to SARSA.  


**Results/Figures:** 

1. Yes, you're completely right. After looking back at my code, I realized I made a mistake in part of my code which I used to log the state, action, reward values to plot the trajectory. This has been addressed in the new code. Also, note that my preliminary results for SARSA and Q-Learning were with only few number of episodes, and steps. That's why my optimal policy results from these two algorithms were different from what I had from PI and VI. I increased the number of my episodes and steps to higher numbers, and tweaked my hyperparameters for the former algorithms to get a result closer to PI and VI. 

2. That's a good idea. I applied this change in my final code and results. 

3. Yes you're right. I have included the correct figures for the pendulum problem in my final report. 

4. Yes, correct. I have applied this change in my final code. 


Thanks for the great feedback!
Mahshid
